lib/ansible/plugins/connection/aws_ssm.py

```diff
@@ -0,0 +1,535 @@
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        ''' run a command on the ssm host '''
+
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvv(u"EXEC {0}".format(to_text(cmd)), host=self.host)
+
+        session = self._session
+
+        mark_start = "".join([random.choice(string.ascii_letters) for i in xrange(self.MARK_LENGTH)])
+        mark_end = "".join([random.choice(string.ascii_letters) for i in xrange(self.MARK_LENGTH)])
+
+        # Wrap command in markers accordingly for the shell used
+        cmd = self._wrap_command(cmd, sudoable, mark_start, mark_end)
+
+        self._flush_stderr(session)
+
+        # Handle the back-end throttling
+        for c in cmd:
+            session.stdin.write(c.encode('utf-8'))
+            if self.is_windows:
+                time.sleep(15 / 1000.0)
+
+        # Read stdout between the markers
+        stdout = ''
+        begin = False
+        stop_time = int(round(time.time())) + self.get_option('timeout')
+        while session.poll() is None:
+            remaining = stop_time - int(round(time.time()))
+            if remaining < 1:
+                self._timeout = True
+                display.vvvv(u"EXEC timeout stdout: {0}".format(to_text(stdout)), host=self.host)
+                raise AnsibleConnectionFailure("SSM exec_command timeout on host: %s"
+                                               % self.instance_id)
+            if self._poll_stdout.poll(1000):
+                line = self._filter_ansi(self._stdout.readline())
+                display.vvvv(u"EXEC stdout line: {0}".format(to_text(line)), host=self.host)
+            else:
+                display.vvvv(u"EXEC remaining: {0}".format(remaining), host=self.host)
+                continue
+
+            if mark_start in line:
+                begin = True
+                if not line.startswith(mark_start):
+                    stdout = ''
+                continue
+            if begin:
+                if mark_end in line:
+                    for line in stdout.splitlines():
+                        display.vvvvv(u"POST_PROCESS: {0}".format(to_text(line)), host=self.host)
+                    returncode, stdout = self._post_process(stdout)
+                    break
+                else:
+                    stdout = stdout + line
+
+        stderr = self._flush_stderr(session)
+
+        return (returncode, stdout, stderr)
+
+    def _prepare_terminal(self):
+        ''' perform any one-time terminal settings '''
+
+        if not self.is_windows:
+            cmd = "stty -echo\n" + "PS1=''\n"
+            if PY3:
+                cmd = to_bytes(cmd)
+            self._session.stdin.write(cmd)
+
+    def _wrap_command(self, cmd, sudoable, mark_start, mark_end):
+        ''' wrap command so stdout and status can be extracted '''
+
+        if self.is_windows:
+            cmd = cmd + "; echo " + mark_start + " $? $LASTEXITCODE\necho " + mark_end + "\n"
```

Comment:
There are 2 things we need to try and do here to make this better at handling errors in PowerShell. We should first add the following conditional

```
# requires the following import
from ansible.plugins.shell.powershell import _common_args

if not cmd.startswith(" ".join(_common_args) + " -EncodedCommand"):
    cmd = self._shell._encode_script(cmd, preserve_rc=True)
```

What this does is detects if `cmd` is already "encoded" by the PowerShell plugin and encodes it if it isn't. This basically makes sure any command that doesn't start with `PowerShell -EncodedCommand ...` is in that format. We do this for 2 reasons

1. We don't have to worry about escaping quotes or any other special chars if we are just running a powershell command
2. The value of `$LASTEXITCODE` should be set to the execution status of just `cmd` and nothing else in the command we generate

Number 2 is important as during testing I found the `PUT` operation was failing on the remote side as `Invoke-WebRequest` was failing but `$LASTEXITCODE` was not set because we weren't running it in a separate process. The downside to this is we are relying on a `_` prefixed function and variable in `powershell.py` and unfortunately I don't know of a better way around this without duplicating code.

This leads into the 2nd point about this line, because you do `echo mark_start`, the [$?](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-6#section-1) variable is set to the status of the last executed command which is the echo. This is useless here and we should just have `$LASTEXITCODE` to detect the output of the last command.

Reply:
@hanumantharaomvl Got your irc message, yes this is the comment I was referring to.

Reply:
Both `$?` and `$LASTEXITCODE` are needed since commands (like `pwd`) do not change the value of `$LASTEXITCODE`.  So in the following example the last command (`pwd`) would look incorrectly like it failed if only `$LASTEXITCODE` was used: 

```
PS C:\Windows\system32> cmd /c exit 99
PS C:\Windows\system32> echo $LASTEXITCODE
99
PS C:\Windows\system32> pwd
Path
----
C:\Windows\system32
PS C:\Windows\system32> echo $LASTEXITCODE
99
PS C:\Windows\system32>
```

The 2nd point here is incorrect in assuming somehow the status of the echo command is being used somehow. Here's a simple test to demonstrate what's actually happening here:

```
PS C:\Windows\system32> cmd /c exit 99
PS C:\Windows\system32> echo start $? $LASTEXITCODE
start
False
99
PS C:\Windows\system32>
```

Reply:
But the major problem with `$?` is that is is only the status of the last executed command as per [the docs for $?](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-6#section-1)

> Contains the execution status of the last command. It contains True if the last command succeeded and False if it failed.

I was finding during testing that the PUT operations for sending a file to the remote was failing but `$?` was returning true so the error was just ignored and then it subsequently failed at a further task. Using the `_encode_script` with `preserve_rc=True` meant that the failed PUT operation was able to detect the actual failure. In any case wrapping the command means

* You now longer have to worry about escaping quotes now and in the future
* You can simplify the error check by just looking at $LASTEXITCODE
* If for some reason the default shell changes to PowerShell Core this forces the module code to run with the tested Windows PowerShell as we invoke `powershell.exe`. This could potentially be user configurable in the future once we support PowerShell Core

Reply:
@jborean93 https://github.com/ansible/ansible/pull/49652/commits/46494097a73fb1327002a6846c33383c13887e1c implements the conditional encoding. Can you mark this as resolved?
One of the tests (https://app.shippable.com/github/ansible/ansible/runs/154989/118/console) failed but this looks like a glitch in Shippable as it never really started doing anything and eventually failed due to a timeout.

Reply:
It looks like there may be an issue with the tests. I've restarted the failed test.

Reply:
Thanks @mattclay!

Reply:
@jborean93 , We have updated the code as you suggested in the commit https://github.com/ansible/ansible/pull/49652/commits/741623e3295a5f6af448da25855e1f13149b7676 and all the checks were passed. Can you please review and approve ?